packages/assets-controllers/src/TokenRatesController.ts

```diff
@@ -416,22 +444,71 @@ export class TokenRatesController extends BaseController<
     if (this.tokenList.length === 0 || this.disabled) {
       return;
     }
-    const slug = await this.getChainSlug();
+
+    const { chainId, nativeCurrency } = this.config;
+    const tokenAddresses = this.tokenList.map((token) => token.address);
+    await this.updateExchangeRatesByChainId({
+      chainId,
+      nativeCurrency,
+      tokenAddresses,
+    });
+
+    this.update({
+      contractExchangeRates:
+        this.state.contractExchangeRatesByChainId[chainId][nativeCurrency],
+    });
+  }
+
+  /**
+   * Updates exchange rates for all tokens.
+   *
+   * @param options - The options to fetch exchange rates.
+   * @param options.chainId - The chain ID.
+   * @param options.nativeCurrency - The ticker for the chain.
+   * @param options.tokenAddresses - The addresses for the token contracts.
+   */
+  async updateExchangeRatesByChainId({
+    chainId,
+    nativeCurrency,
+    tokenAddresses,
+  }: {
+    chainId: Hex;
+    nativeCurrency: string;
+    tokenAddresses: string[];
+  }) {
+    if (!tokenAddresses.length) {
+      return;
+    }
+    const chainSlug = await this.getChainSlug(chainId);
 
     let newContractExchangeRates: ContractExchangeRates = {};
-    if (!slug) {
-      this.tokenList.forEach((token) => {
-        const address = toChecksumHexAddress(token.address);
+    if (!chainSlug) {
+      tokenAddresses.forEach((tokenAddress) => {
+        const address = toChecksumHexAddress(tokenAddress);
         newContractExchangeRates[address] = undefined;
       });
     } else {
-      const { nativeCurrency } = this.config;
       newContractExchangeRates = await this.fetchAndMapExchangeRates(
         nativeCurrency,
-        slug,
+        chainSlug,
+        tokenAddresses,
       );
     }
-    this.update({ contractExchangeRates: newContractExchangeRates });
+
+    const existingContractExchangeRates =
```

Comment:
this name got me confused thinking it was `this.state.contractExchangeRates`. Maybe it would be better to make it `existingContractExchangeRatesForChainId`?

Reply:
done https://github.com/MetaMask/core/pull/2015/commits/d460c18b98b7a24a0cb4902aa43369ab3dbf647c